src/components/tl/mlx5/mcast/tl_mlx5_mcast.h

```diff
@@ -122,19 +311,63 @@ typedef struct ucc_tl_mlx5_mcast_oob_p2p_context {
     ucc_subset_t  subset;
 } ucc_tl_mlx5_mcast_oob_p2p_context_t;
 
-#define TASK_TEAM_MCAST(_task)                                                       \
-    (ucc_derived_of((_task)->super.team, ucc_tl_mlx5_mcast_team_t))
-#define TASK_CTX_MCAST(_task)                                                        \
-    (ucc_derived_of((_task)->super.team->context, ucc_tl_mlx5_mcast_context_t))
-#define TASK_LIB_MCAST(_task)                                                        \
-    (ucc_derived_of((_task)->super.team->context->lib, ucc_tl_mlx5_mcast_lib_t))
+static inline struct pp_packet* ucc_tl_mlx5_mcast_buf_get_free(ucc_tl_mlx5_mcast_coll_comm_t* comm)
+{
+    struct pp_packet* pp;
+    
+    pp = ucc_list_extract_head(&comm->bpool, struct pp_packet, super);
+
+    ucc_assert(pp == NULL || pp->context == 0);
+
+    return pp;
+}
+
+static inline ucc_status_t ucc_tl_mlx5_mcast_post_recv_buffers(ucc_tl_mlx5_mcast_coll_comm_t* comm)
+{
+    struct ibv_recv_wr *bad_wr = NULL;
+    struct ibv_recv_wr *rwr    = comm->call_rwr;
+    struct ibv_sge     *sge    = comm->call_rsgs;
+    struct pp_packet   *pp     = NULL;
+    int                 i;
+    int                 count;
+    
+    count = comm->params.rx_depth - comm->pending_recv;
+
+    if (count <= comm->params.post_recv_thresh) {
+        return UCC_OK;
+    }
+
+    for (i = 0; i < count; i++) {
+        if (NULL == (pp = ucc_tl_mlx5_mcast_buf_get_free(comm))) {
+            break;
+        }
+
+        rwr[i].wr_id      = ((uint64_t) pp);
+        rwr[i].next       = &rwr[i+1];
+        sge[2*i + 1].addr = pp->buf;
+
+        ucc_assert((uint64_t)comm->pp <= rwr[i].wr_id
+                && ((uint64_t)comm->pp + comm->buf_n * sizeof(struct pp_packet)) > rwr[i].wr_id);
+    }
+    if (i > 0) {
+        rwr[i-1].next = NULL;
+        if (ibv_post_recv(comm->mcast.qp, &rwr[0], &bad_wr)) {
+            tl_error(comm->lib, "failed to prepost recvs: errno %d", errno);
+            return UCC_ERR_NO_RESOURCE;
+        }
+        comm->pending_recv += i;
+    }
+
+    return UCC_OK;
+}
+
 #define TASK_ARGS_MCAST(_task) (_task)->super.bargs.args
```

Comment:
looks like TASK_ARGS definition has changed since I made the code. Now it is same as _MCAST version. Therefore, there is no need for TASK_ARGS_MCAST anymore. I will remove it.